File: wizards/com/sun/star/wizards/ui/ControlScroller.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */
package com.sun.star.wizards.ui;

import com.sun.star.beans.*;
import com.sun.star.awt.*;
import com.sun.star.wizards.common.*;

import java.util.*;

public abstract class ControlScroller
{

    protected WizardDialog CurUnoDialog;
    protected int ncurfieldcount;
    private int nblockincrement;
    private int nlineincrement;
    private int nscrollvalue = 0;
    private int ntotfieldcount;
    private XScrollBar xScrollBar;
    private ArrayList<PropertyValue[]> scrollfields;
    protected int iCompPosX;
    protected int iCompPosY;
    protected int iCompWidth;
    protected int iCompHeight;
    protected short curtabindex;
    protected Integer IStep;
    int iScrollBarWidth = 10;
    private int SORELFIRSTPOSY = 3;
    protected int curHelpIndex;
    private String sIncSuffix;
    protected ArrayList<Object> ControlGroupVector = new ArrayList<Object>();

    private class AdjustmentListenerImpl implements com.sun.star.awt.XAdjustmentListener
    {

        public void disposing(com.sun.star.lang.EventObject eventObject)
        {
        }

        public void adjustmentValueChanged(AdjustmentEvent AdjustEvent)
        {
            scrollControls();
        }
    }

    /**
     *
     * @param _CurUnoDialog
     * @param _iStep
     * @param _iCompPosX
     * @param _iCompPosY
     * @param _iCompWidth
     * @param _nblockincrement
     * @param _firsthelpindex
     */
    // TODO add parameters for tabindices and helpindex
    protected ControlScroller(WizardDialog _CurUnoDialog, int _iStep, int _iCompPosX, int _iCompPosY, int _iCompWidth, int _nblockincrement, int _nlinedistance, int _firsthelpindex)
    {
        this.nblockincrement = _nblockincrement;
        this.CurUnoDialog = _CurUnoDialog;
        this.curHelpIndex = _firsthelpindex;
        curtabindex = UnoDialog.setInitialTabindex(_iStep);
        IStep = Integer.valueOf(_iStep);
        this.iCompPosX = _iCompPosX;
        this.iCompPosY = _iCompPosY;
        this.iCompWidth = _iCompWidth;
        this.iCompHeight = 2 * SORELFIRSTPOSY + nblockincrement * _nlinedistance;
        int iStartPosY = iCompPosY + SORELFIRSTPOSY;
        int ScrollHeight = iCompHeight - 2;
        nlineincrement = 1;
        sIncSuffix = com.sun.star.wizards.common.Desktop.getIncrementSuffix(CurUnoDialog.getDlgNameAccess(), "TitleScrollBar");
        xScrollBar = CurUnoDialog.insertScrollBar("TitleScrollBar" + sIncSuffix, 0,
                new AdjustmentListenerImpl(),
                new String[]
                {
                    PropertyNames.PROPERTY_BORDER, PropertyNames.PROPERTY_ENABLED, PropertyNames.PROPERTY_HEIGHT, PropertyNames.PROPERTY_HELPURL, PropertyNames.ORIENTATION, PropertyNames.PROPERTY_POSITION_X, PropertyNames.PROPERTY_POSITION_Y, PropertyNames.PROPERTY_STEP, PropertyNames.PROPERTY_WIDTH
                },
                new Object[]
                {
                    Short.valueOf((short) 0), Boolean.TRUE, Integer.valueOf(ScrollHeight), HelpIds.getHelpIdString(curHelpIndex), Integer.valueOf(ScrollBarOrientation.VERTICAL), Integer.valueOf(iCompPosX + iCompWidth - iScrollBarWidth - 1), Integer.valueOf(iCompPosY + 1), IStep, Integer.valueOf(iScrollBarWidth)
                });
        scrollfields = new ArrayList<PropertyValue[]>();
        int ypos = iStartPosY + SORELFIRSTPOSY;
        for (int i = 0; i < nblockincrement; i++)
        {
            insertControlGroup(i, ypos);
            ypos += _nlinedistance;
        }
    }



    /**
     * @param _ntotfieldcount The number of fields that are to be administered by the ControlScroller
     */
    protected void initialize(int _ntotfieldcount)
    {
        try
        {
            ntotfieldcount = _ntotfieldcount;
            setCurFieldCount();
            nscrollvalue = 0;
            Helper.setUnoPropertyValue(UnoDialog.getModel(xScrollBar), "ScrollValue", Integer.valueOf(nscrollvalue));
            if (ntotfieldcount > nblockincrement)
            {
                Helper.setUnoPropertyValues(UnoDialog.getModel(xScrollBar), new String[]
                        {
                            PropertyNames.PROPERTY_ENABLED, "BlockIncrement", "LineIncrement", "ScrollValue", "ScrollValueMax"
                        }, new Object[]
                        {
                            Boolean.TRUE, Integer.valueOf(nblockincrement), Integer.valueOf(nlineincrement), Integer.valueOf(nscrollvalue), Integer.valueOf(ntotfieldcount - nblockincrement)
                        });
            }
            else
            {
                Helper.setUnoPropertyValues(UnoDialog.getModel(xScrollBar), new String[]
                        {
                            PropertyNames.PROPERTY_ENABLED, "ScrollValue"
                        }, new Object[]
                        {
                            Boolean.FALSE, Integer.valueOf(nscrollvalue)
                        });
            }
            fillupControls(true);
        }
        catch (java.lang.Exception ex)
        {
            ex.printStackTrace();
        }
    }

    private void fillupControls(boolean binitialize)
    {
        for (int a = 0; a < this.nblockincrement; a++)
        {
            if (a < ncurfieldcount)
            {
                fillupControls(a);
            }
            if (binitialize)
            {
                setControlGroupVisible(a, (a < this.ncurfieldcount));
            }
        }
        if (binitialize)
        {
            CurUnoDialog.repaintDialogStep();
        }
    }

    private void fillupControls(int guiRow)
    {
        PropertyValue[] nameProps = scrollfields.get(guiRow);
        PropertyValue[] valueProps = scrollfields.get(guiRow + nscrollvalue);
        for (int n = 0; n < nameProps.length; n++)
        {
            if (CurUnoDialog.getDlgNameAccess().hasByName(nameProps[n].Name))
            {
                setControlData(nameProps[n].Name, valueProps[n].Value);
            }
            else
            {
                throw new IllegalArgumentException("No such control !");
            }
        }
    }

    private void setScrollValue(int _nscrollvalue)
    {
        if (_nscrollvalue >= 0)
        {
            Helper.setUnoPropertyValue(UnoDialog.getModel(xScrollBar), "ScrollValue", Integer.valueOf(_nscrollvalue));
            scrollControls();
        }
    }

    protected void setScrollValue(int _nscrollvalue, int _ntotfieldcount)
    {
        setTotalFieldCount(_ntotfieldcount);
        setScrollValue(_nscrollvalue);
    }

    protected int getTotalFieldCount()
    {
        return ntotfieldcount;
    }

    protected int getCurFieldCount()
    {
        return ncurfieldcount;
    }

    private void setCurFieldCount()
    {
        if (ntotfieldcount > nblockincrement)
        {
            ncurfieldcount = nblockincrement;
        }
        else
        {
            ncurfieldcount = ntotfieldcount;
        }
    }

    protected void setTotalFieldCount(int _ntotfieldcount)
    {
        this.ntotfieldcount = _ntotfieldcount;
        setCurFieldCount();
        if (ntotfieldcount > nblockincrement)
        {
            Helper.setUnoPropertyValues(UnoDialog.getModel(xScrollBar), new String[]
                    {
                        PropertyNames.PROPERTY_ENABLED, "ScrollValueMax"
                    }, new Object[]
                    {
                        Boolean.TRUE, Integer.valueOf(ntotfieldcount - nblockincrement)
                    });
        }
        else
        {
            Helper.setUnoPropertyValue(UnoDialog.getModel(xScrollBar), PropertyNames.PROPERTY_ENABLED, Boolean.FALSE);
        }
    }

    protected void toggleComponent(boolean _bdoenable)
    {
        boolean bdoenable = _bdoenable && (ntotfieldcount > nblockincrement);
        CurUnoDialog.setControlProperty("TitleScrollBar" + sIncSuffix, PropertyNames.PROPERTY_ENABLED, Boolean.valueOf(bdoenable));
    }

    protected void toggleControls(boolean _bdoenable)
    {
        for (int n = 0; n < scrollfields.size(); n++)
        {
            PropertyValue[] curproperties = scrollfields.get(n);
            for (int m = 0; m < curproperties.length; m++)
            {
                PropertyValue curproperty = curproperties[m];
                CurUnoDialog.setControlProperty(curproperty.Name, PropertyNames.PROPERTY_ENABLED, Boolean.valueOf(_bdoenable));
            }
        }

    }

    protected int getScrollValue()
    {
        return nscrollvalue;
    }







    protected int getBlockIncrementation()
    {
        return nblockincrement;
    }

    private void scrollControls()
    {
        try
        {
            scrollRowsInfo();
            nscrollvalue = ((Integer) Helper.getUnoPropertyValue(UnoDialog.getModel(xScrollBar), "ScrollValue")).intValue();
            if (nscrollvalue + nblockincrement >= ntotfieldcount)
            {
                nscrollvalue = (ntotfieldcount) - nblockincrement;
            }
            fillupControls(false);
        }
        catch (java.lang.Exception ex)
        {
            ex.printStackTrace();
        }
    }

    private void scrollRowsInfo()
    {
        int cols =
                scrollfields.size() > 0
                ? scrollfields.get(0).length
                : 0;
        for (int a = 0; a < ncurfieldcount; a++)
        {
            for (int n = 0; n < cols; n++)
            {
                fieldInfo(a, n);
            }
        }
    }

    /**
     * updates the corresponding data to
     * the control in guiRow and column
     * @param guiRow 0 based row index
     * @param column 0 based column index
     * @return the propertyValue object corresponding to
     * this control.
     */
    private PropertyValue fieldInfo(int guiRow, int column)
    {
        if (guiRow + nscrollvalue < scrollfields.size())
        {
            return fieldInfo(
                    scrollfields.get(guiRow + nscrollvalue)[column],
                    scrollfields.get(guiRow)[column]);
        }
        else
        {
            return null;
        }
    }

    private PropertyValue fieldInfo(PropertyValue valueProp, PropertyValue nameProp)
    {
        if (CurUnoDialog.getDlgNameAccess().hasByName(nameProp.Name))
        {
            valueProp.Value = getControlData(nameProp.Name);
        }
        else
        {
            valueProp.Value = nameProp.Value;
        }
        return valueProp;
    }

    protected void unregisterControlGroup(int _index)
    {
        scrollfields.remove(_index);
    }

    protected void registerControlGroup(PropertyValue[] _currowproperties, int _i)
    {
        if (_i == 0)
        {
            scrollfields.clear();
        }
        if (_i >= scrollfields.size())
        {
            scrollfields.add(_currowproperties);
        }
        else
        {
            scrollfields.set(_i, _currowproperties);
        }
    }



    private void setControlData(String controlname, Object newvalue)
    {
        Object oControlModel = UnoDialog.getModel(CurUnoDialog.xDlgContainer.getControl(controlname));
        String propertyname = UnoDialog.getDisplayProperty(oControlModel);
        if (!propertyname.equals(PropertyNames.EMPTY_STRING))
        {
            CurUnoDialog.setControlProperty(controlname, propertyname, newvalue);
        }
    }

    private Object getControlData(String controlname)
    {
        Object oControlModel = UnoDialog.getModel(CurUnoDialog.xDlgContainer.getControl(controlname));
        String propertyname = UnoDialog.getDisplayProperty(oControlModel);
        if (!propertyname.equals(PropertyNames.EMPTY_STRING))
        {
            return CurUnoDialog.getControlProperty(controlname, propertyname);
        }
        else
        {
            return null;
        }
    }

    protected PropertyValue[][] getScrollFieldValues()
    {
        scrollRowsInfo();
        PropertyValue[] curproperties;
        PropertyValue[][] retproperties;
        retproperties = new PropertyValue[scrollfields.size()][];
        try
        {
            for (int i = 0; i < scrollfields.size(); i++)
            {
                curproperties = scrollfields.get(i);
                retproperties[i] = curproperties;
            }
            return retproperties;
        }
        catch (java.lang.Exception ex)
        {
            ex.printStackTrace(System.err);
            return null;
        }
    }

    protected abstract void initializeScrollFields();

    /** inserts a group of controls into the component. The group may either be a row or a column of controls
     *  The controls should be put on Step 99 (means made invisible at first). All the controlrows that are needed are than
     *  made visible automatically when calling "initialize(_fieldcount)"
     * @param _index The index of the control group
     * @param npos Can be an x coordinate or an y coordinate which depends on the orientation of the scrollbar
     */
    protected abstract void insertControlGroup(int _index, int npos);

    protected abstract void setControlGroupVisible(int _index, boolean _bIsVisible);
}

File: wizards/com/sun/star/wizards/ui/ControlScroller.py
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# This file incorporates work covered by the following license notice:
#
#   Licensed to the Apache Software Foundation (ASF) under one or more
#   contributor license agreements. See the NOTICE file distributed
#   with this work for additional information regarding copyright
#   ownership. The ASF licenses this file to you under the Apache
#   License, Version 2.0 (the "License"); you may not use this file
#   except in compliance with the License. You may obtain a copy of
#   the License at http://www.apache.org/licenses/LICENSE-2.0 .
#
import uno
import traceback
from wizards.common.Desktop import Desktop
from wizards.common.PropertyNames import PropertyNames
from wizards.common.HelpIds import HelpIds
from wizards.common.Helper import Helper
from wizards.ui.UnoDialog import UnoDialog

from com.sun.star.awt.ScrollBarOrientation import HORIZONTAL, VERTICAL

class ControlScroller(object):

    SORELFIRSTPOSY = 3
    iScrollBarWidth = 10
    scrollfields = []
    CurUnoDialog = None
    iStep = None
    curHelpIndex = None

    # TODO add parameters for tabindices and helpindex
    def __init__(self, _CurUnoDialog, _xMSF, _iStep, _iCompPosX, _iCompPosY,
            _iCompWidth, _nblockincrement, _nlinedistance, _firsthelpindex):
        self.xMSF = _xMSF
        ControlScroller.nblockincrement = _nblockincrement
        ControlScroller.CurUnoDialog = _CurUnoDialog
        ControlScroller.iStep = _iStep
        ControlScroller.curHelpIndex = _firsthelpindex
        self.curtabindex = ControlScroller.iStep * 100
        self.linedistance = _nlinedistance
        self.iCompPosX = _iCompPosX
        self.iCompPosY = _iCompPosY
        self.iCompWidth = _iCompWidth
        self.iCompHeight = 2 * ControlScroller.SORELFIRSTPOSY + \
                ControlScroller.nblockincrement * self.linedistance
        self.iStartPosY = self.iCompPosY + ControlScroller.SORELFIRSTPOSY
        ScrollHeight = self.iCompHeight - 2
        self.nlineincrement = 1
        self.sincSuffix = Desktop.getIncrementSuffix(
            ControlScroller.CurUnoDialog.xDialogModel, "imgBackground")
        self.oImgControl = ControlScroller.CurUnoDialog.insertControlModel(
            "com.sun.star.awt.UnoControlImageControlModel",
            "imgBackground" + self.sincSuffix,
            ("Border", PropertyNames.PROPERTY_HEIGHT,
                PropertyNames.PROPERTY_POSITION_X,
                PropertyNames.PROPERTY_POSITION_Y,
                PropertyNames.PROPERTY_STEP,
                PropertyNames.PROPERTY_WIDTH),
            (1, self.iCompHeight, self.iCompPosX,
                self.iCompPosY, ControlScroller.iStep, self.iCompWidth))
        self.oImgControl = ControlScroller.CurUnoDialog.xUnoDialog.getControl(
            "imgBackground" + self.sincSuffix)
        self.setComponentMouseTransparent()
        ControlScroller.xScrollBar = ControlScroller.CurUnoDialog.insertScrollBar(
            "TitleScrollBar" + self.sincSuffix,
            ("Border", PropertyNames.PROPERTY_ENABLED,
                PropertyNames.PROPERTY_HEIGHT,
                PropertyNames.PROPERTY_HELPURL, "Orientation",
                PropertyNames.PROPERTY_POSITION_X,
                PropertyNames.PROPERTY_POSITION_Y,
                PropertyNames.PROPERTY_STEP,
                PropertyNames.PROPERTY_WIDTH),
            (0, True, ScrollHeight,
                HelpIds.getHelpIdString(ControlScroller.curHelpIndex),
                VERTICAL, self.iCompPosX + self.iCompWidth - \
                    ControlScroller.iScrollBarWidth - 1,
                self.iCompPosY + 1, ControlScroller.iStep,
                ControlScroller.iScrollBarWidth), 0, self)
        ControlScroller.nscrollvalue = 0
        ControlScroller.ControlGroupVector = []
        ypos = self.iStartPosY + ControlScroller.SORELFIRSTPOSY
        for i in xrange(ControlScroller.nblockincrement):
            self.insertControlGroup(i, ypos)
            ypos += self.linedistance

    def setComponentMouseTransparent(self):
        ControlScroller.CurUnoDialog.getPeerConfiguration().\
            setPeerProperties(self.oImgControl, "MouseTransparent", True)

    def setScrollBarOrientationHorizontal(self):
        Helper.setUnoPropertyValue(
            ControlScroller.xScrollBar, "Orientation", HORIZONTAL)

    @classmethod
    def fillupControls(self, binitialize):
        for i in xrange(ControlScroller.nblockincrement):
            if i < self.ncurfieldcount:
                self.fillupControl(i)

        if binitialize:
            ControlScroller.CurUnoDialog.repaintDialogStep()

    @classmethod
    def fillupControl(self, guiRow):
        nameProps = ControlScroller.scrollfields[guiRow]
        valueProps = ControlScroller.scrollfields[guiRow + ControlScroller.nscrollvalue]
        for index, item in enumerate(nameProps):
            if ControlScroller.CurUnoDialog.xDialogModel.hasByName(item.Name):
                self.setControlData(item.Name, valueProps[index].Value)
            else:
                raise AttributeError("No such control !")

    @classmethod
    def setScrollValue(self, _nscrollvalue, _ntotfieldcount=None):
        if _ntotfieldcount is not None:
            self.setTotalFieldCount(_ntotfieldcount)
        if _nscrollvalue >= 0:
            Helper.setUnoPropertyValue(
                ControlScroller.xScrollBar.Model, "ScrollValue", _nscrollvalue)
            self.scrollControls()

    @classmethod
    def setCurFieldCount(self):
        if self.ntotfieldcount > ControlScroller.nblockincrement:
            self.ncurfieldcount = ControlScroller.nblockincrement
        else:
            self.ncurfieldcount = self.ntotfieldcount

    @classmethod
    def setTotalFieldCount(self, _ntotfieldcount):
        self.ntotfieldcount = _ntotfieldcount
        self.setCurFieldCount()
        if self.ntotfieldcount > ControlScroller.nblockincrement:
            Helper.setUnoPropertyValues(
                ControlScroller.xScrollBar.Model,
                (PropertyNames.PROPERTY_ENABLED, "ScrollValueMax"),
                (True, self.ntotfieldcount - ControlScroller.nblockincrement))
        else:
            Helper.setUnoPropertyValue(ControlScroller.xScrollBar.Model,
                PropertyNames.PROPERTY_ENABLED, False)

    def toggleComponent(self, _bdoenable):
        bdoenable = _bdoenable and \
            (self.ntotfieldcount > ControlScroller.nblockincrement)
        ControlScroller.CurUnoDialog.setControlProperty(
                "TitleScrollBar" + self.sincSuffix,
                PropertyNames.PROPERTY_ENABLED, bdoenable)

    def toggleControls(self, _bdoenable):
        n = 0
        while n < ControlScroller.scrollfields.size():
            curproperties = ControlScroller.scrollfields.elementAt(n)
            m = 0
            while m < curproperties.length:
                curproperty = curproperties[m]
                ControlScroller.CurUnoDialog.setControlProperty(
                curproperty.Name, PropertyNames.PROPERTY_ENABLED, _bdoenable)
                m += 1
            n += 1

    @classmethod
    def scrollControls(self):
        try:
            ControlScroller.nscrollvalue = int(Helper.getUnoPropertyValue(
                    ControlScroller.xScrollBar.Model, "ScrollValue"))
            if ControlScroller.nscrollvalue + ControlScroller.nblockincrement \
                    >= self.ntotfieldcount:
                ControlScroller.nscrollvalue = \
                    self.ntotfieldcount - ControlScroller.nblockincrement
            self.fillupControls(False)
        except Exception:
            traceback.print_exc()

    '''
    updates the corresponding data to
    the control in guiRow and column
    @param guiRow 0 based row index
    @param column 0 based column index
    @return the propertyValue object corresponding to
    this control.
    '''

    @classmethod
    def fieldInfo(self, guiRow, column):
        if guiRow + ControlScroller.nscrollvalue < len(ControlScroller.scrollfields):
            valueProp = (ControlScroller.scrollfields[guiRow + ControlScroller.nscrollvalue])[column]
            nameProp = (ControlScroller.scrollfields[guiRow])[column]
            if ControlScroller.CurUnoDialog.xDialogModel.hasByName(nameProp.Name):
                valueProp.Value = self.getControlData(nameProp.Name)
            else:
                valueProp.Value = nameProp.Value
            return valueProp
        else:
            return None

    @classmethod
    def unregisterControlGroup(self, _index):
        del ControlScroller.scrollfields[_index]

    @classmethod
    def registerControlGroup(self, _currowproperties, _i):
        if _i == 0:
            del ControlScroller.scrollfields[:]

        if _i >= len(ControlScroller.scrollfields):
            ControlScroller.scrollfields.append(_currowproperties)
        else:
            ControlScroller.scrollfields.insert(_currowproperties, _i)

    @classmethod
    def setControlData(self, controlname, newvalue):
        oControlModel = ControlScroller.CurUnoDialog.xUnoDialog.getControl(
            controlname).Model
        propertyname = UnoDialog.getDisplayProperty(oControlModel)
        if propertyname != "":
            ControlScroller.CurUnoDialog.setControlProperty(
                controlname, propertyname, newvalue)

    @classmethod
    def getControlData(self, controlname):
        oControlModel = ControlScroller.CurUnoDialog.xUnoDialog.getControl(
            controlname).Model
        propertyname = UnoDialog.getDisplayProperty(oControlModel)
        if propertyname != "":
            return ControlScroller.CurUnoDialog.getControlProperty(
                controlname, propertyname)
        else:
            return None

    def getScrollFieldValues(self):
        retproperties = [[ControlScroller.scrollfields.size()],[]]
        try:
            i = 0
            while i < ControlScroller.scrollfields.size():
                curproperties = ControlScroller.scrollfields.elementAt(i)
                retproperties[i] = curproperties
                i += 1
            return retproperties
        except java.lang.Exception, ex:
            ex.printStackTrace(System.out)
            return None
